chia/util/condition_tools.py

```diff
@@ -55,41 +54,157 @@ def parse_sexp_to_conditions(sexp: Program) -> List[ConditionWithArgs]:
+                msg + spend.parent_id + int_to_bytes(spend.coin_amount) + data[ConditionOpcode.AGG_SIG_PARENT_AMOUNT]
+            )
+
+        for pk, msg in spend.agg_sig_parent_puzzle:
+            ret[0].append(bytes48(pk))
+            ret[1].append(msg + spend.parent_id + spend.puzzle_hash + data[ConditionOpcode.AGG_SIG_PARENT_PUZZLE])
+
         for pk, msg in spend.agg_sig_me:
             ret[0].append(bytes48(pk))
-            ret[1].append(msg + spend.coin_id + additional_data)
+            ret[1].append(msg + spend.coin_id + data[ConditionOpcode.AGG_SIG_ME])
+
     return ret
 
 
+def validate_cwa(cwa: ConditionWithArgs) -> None:
+    if (
+        len(cwa.vars) != 2
+        or len(cwa.vars[0]) != 48
+        or len(cwa.vars[1]) > 1024
+        or cwa.vars[0] is None
+        or cwa.vars[1] is None
+    ):
+        raise ConsensusError(Err.INVALID_CONDITION)
+
+
 def pkm_pairs_for_conditions_dict(
-    conditions_dict: Dict[ConditionOpcode, List[ConditionWithArgs]], coin_name: bytes32, additional_data: bytes
+    conditions_dict: Dict[ConditionOpcode, List[ConditionWithArgs]],
+    coin: Coin,
+    additional_data: bytes,
 ) -> List[Tuple[bytes48, bytes]]:
-    assert coin_name is not None
     ret: List[Tuple[bytes48, bytes]] = []
 
+    data = agg_sig_additional_data(additional_data)
+
     for cwa in conditions_dict.get(ConditionOpcode.AGG_SIG_UNSAFE, []):
-        assert len(cwa.vars) == 2
-        assert len(cwa.vars[0]) == 48 and len(cwa.vars[1]) <= 1024
-        assert cwa.vars[0] is not None and cwa.vars[1] is not None
-        if cwa.vars[1].endswith(additional_data):
-            raise ConsensusError(Err.INVALID_CONDITION)
+        validate_cwa(cwa)
+        for disallowed in data.values():
+            if cwa.vars[1].endswith(disallowed):
+                raise ConsensusError(Err.INVALID_CONDITION)
         ret.append((bytes48(cwa.vars[0]), cwa.vars[1]))
 
+    for cwa in conditions_dict.get(ConditionOpcode.AGG_SIG_PARENT, []):
+        validate_cwa(cwa)
+        ret.append(
+            (
+                bytes48(cwa.vars[0]),
+                cwa.vars[1] + coin.parent_coin_info + data[ConditionOpcode.AGG_SIG_PARENT],
+            )
+        )
+
+    for cwa in conditions_dict.get(ConditionOpcode.AGG_SIG_PUZZLE, []):
+        validate_cwa(cwa)
+        ret.append((bytes48(cwa.vars[0]), cwa.vars[1] + coin.puzzle_hash + data[ConditionOpcode.AGG_SIG_PUZZLE]))
+
+    for cwa in conditions_dict.get(ConditionOpcode.AGG_SIG_AMOUNT, []):
+        validate_cwa(cwa)
+        ret.append(
+            (
+                bytes48(cwa.vars[0]),
+                cwa.vars[1] + int_to_bytes(coin.amount) + data[ConditionOpcode.AGG_SIG_AMOUNT],
+            )
+        )
+
+    for cwa in conditions_dict.get(ConditionOpcode.AGG_SIG_PUZZLE_AMOUNT, []):
+        validate_cwa(cwa)
+        ret.append(
+            (
+                bytes48(cwa.vars[0]),
+                cwa.vars[1]
+                + coin.puzzle_hash
```

Comment:
Would you consider moving all six of the new `AGG_SIG` types into a loop and using a function parameter to determine how to turn the `coin` into a blob with a `serialized_coin_data_for_aggsig_type` sort of function? It seems a bit curious to have so many loops that are so similar. Very vulnerable to typos.

Reply:
something like

```python
COIN_TO_BLOB_F_LOOKUP: Callable[[Coin], bytes] = {
   ConditionOpcode.AGG_SIG_PARENT: lambda coin: coin.parent_info,
    ConditionOpcode.AGG_SIG_PUZZLE_AMOUNT: lambda coin: coin.puzzle_hash + int_to_bytes(coin.amount),
    # etc
}
```

Reply:
Sure, I sent ac43b34 to that effect.